GVCF "reblocking" code to compress homRef blocks

ldgauthier opened the pull request:
This is a port of the GATK3 version I actually ran.  Some differences in the genotyping engines and the fixed median calculation in MathUtils between GATK3 and GATK4 make the results _slightly_ different in some cases, but still within tolerances.

lbergelson commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/variantutils/ReblockGVCFIntegrationTest.java:
```diff
@@ -0,0 +1,86 @@
+import java.util.Collections;
+
+/**
+ * Created by gauthier on 10/2/17.
```
we typically don't include these since git captures it

ldgauthier replied:
Finally found where to turn that off (Preferences > Editor > File and Code Templates > Includes).

lbergelson commented on src/main/java/org/broadinstitute/hellbender/utils/variant/writers/GVCFWriter.java:
```diff
@@ -155,9 +155,10 @@ public boolean checkError() {
-                return null;
+            //there's a use case here related to ReblockGVCFs for overlapping deletions on different haplotypes
+            if ( vc.getStart() <= nextAvailableStart && vc.getContig().equals(contigOfNextAvailableStart) ) {
+                if (vc.getEnd() <= nextAvailableStart)
```
Could you add a unit case that triggers the first condition but not the second one?

ldgauthier replied:
Yep, I have new tests in GVCFWriterUnitTests.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/ReblockGVCF.java:
```diff
@@ -0,0 +1,453 @@
+        // Remove GCVFBlocks
+        headerLines.removeIf(vcfHeaderLine -> vcfHeaderLine.getKey().startsWith(GVCF_BLOCK));
+
+        headerLines.remove(GATKVCFHeaderLines.getInfoLine(GATKVCFConstants.DOWNSAMPLED_KEY));
```
what happens if someone adds these in annotationsToKeep?

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/ReblockGVCF.java:
```diff
@@ -0,0 +1,453 @@
+     * @param originalVC     the combined genomic VC
+     * @return a new VariantContext or null if the site turned monomorphic and we don't want such sites
+     */
+    protected VariantContext regenotypeVC(final ReferenceContext ref, final VariantContext originalVC) {
```
Is this method named / documented correctly?  It seems more like reblocking them with some minimal stripping out of alleles rather than regenotyping.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/ReblockGVCF.java:
```diff
@@ -0,0 +1,453 @@
+        }
+        //high quality variant
+        else {
+            Map<String, Object> attrMap = new HashMap<>();
```
It would be good to have unit tests for this and reblockVariant, and makeGQ0RefCall.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/ReblockGVCF.java:
```diff
@@ -0,0 +1,453 @@
+    protected boolean dropLowQuals = false;
+
+    @Advanced
+    @Argument(fullName="RGQ-threshold-to-drop", shortName="RGQ-threshold", doc="Reference genotype quality (PL[0]) value below which variant sites will be converted to GQ0 homRef calls")
```
Should this have a min value of 0.0?

ldgauthier replied:
I tried to sneak this one past because I don't like the new style.

The default is zero so that the default output only has homRef blocks modified, but Eric was convinced we could get better compression by throwing out the low quality variants, which is probably true but we didn't do a real test.  The arg name could be more accurate.

ldgauthier commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/ReblockGVCF.java:
```diff
@@ -0,0 +1,453 @@
+    protected List<String> annotationsToUse = Arrays.asList(VCFConstants.DEPTH_KEY, VCFConstants.RMS_MAPPING_QUALITY_KEY, GATKVCFConstants.READ_POS_RANK_SUM_KEY, GATKVCFConstants.MAP_QUAL_RANK_SUM_KEY, GATKVCFConstants.RAW_RMS_MAPPING_QUALITY_KEY);
+
+    @Advanced
+    @Argument(fullName="GVCFGQBands", shortName="GQB", doc="Exclusive upper bounds for reference confidence GQ bands " +
```
I just realized this tool has a HaplotypeCallerArgumentCollection to initialize the GenotypingEngine.  Is there a spiffy way to give GVCFGQBands a different default for this tool or should I just subclass the arg collection?

ldgauthier commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/ReblockGVCF.java:
```diff
@@ -0,0 +1,453 @@
+     */
+    protected VariantContext regenotypeVC(final ReferenceContext ref, final VariantContext originalVC) {
+        if ( originalVC == null ) {
+            throw new IllegalArgumentException("originalVC cannot be null");
```
That's a relief.  I always thought that the GATK3 engine providing null records was embarrassing.

ldgauthier commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/ReblockGVCF.java:
```diff
@@ -0,0 +1,453 @@
+        if(result.getReference().length() > 1) {
+            attrMap.put("END", result.getEnd());
+            newRef = Allele.create(newRef.getBases()[0], true);
+            gb.alleles(Arrays.asList(newRef, newRef));
```
Oooh, that's a nice way to encapsulate the ploidy assumption.

ldgauthier commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/ReblockGVCF.java:
```diff
@@ -0,0 +1,453 @@
+            gb.alleles(Arrays.asList(newRef, newRef));
+        }
+        //there are some cases where there are low quality variants with homRef calls AND alt alleles!
+        //TODO: the best thing would be to take the most likely alt's likelihoods
```
I'm going to pass the buck here.  If we end up running this tool again I'll fix it.

ldgauthier commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/ReblockGVCF.java:
```diff
@@ -0,0 +1,453 @@
+        //there are some cases where there are low quality variants with homRef calls AND alt alleles!
+        //TODO: the best thing would be to take the most likely alt's likelihoods
+        else if (originalVC.getNAlleles() > 2) {
+            final int[] idxVector = originalVC.getGLIndecesOfAlternateAllele(Allele.NON_REF_ALLELE);   //this is always length 3
```
HTSJDK 3 -- now with correct spelling!

ldgauthier commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/ReblockGVCF.java:
```diff
@@ -0,0 +1,453 @@
+                attrMap.put(GATKVCFConstants.RAW_QUAL_APPROX_KEY, result.getGenotype(0).getPL()[0]);
+                int varDP = QualByDepth.getDepth(result.getGenotypes(), null);
+                if (varDP == 0) {  //prevent QD=Infinity case
+                    varDP = result.getAttributeAsInt(VCFConstants.DEPTH_KEY, Integer.MAX_VALUE);
```
I guess that's fair.  My goal was to force QD to zero because people often hard filter on that and clearly this is a weird variant if it's missing all its depth. I really hate that we "correct" high QD values.  It's on my list.

ldgauthier commented on src/main/java/org/broadinstitute/hellbender/utils/variant/GATKVCFHeaderLines.java:
```diff
@@ -14,9 +15,15 @@
-    public static VCFFormatHeaderLine getFormatLine(final String id) { return formatLines.get(id); }
-    public static VCFFilterHeaderLine getFilterLine(final String id) { return filterLines.get(id); }
+    public static VCFInfoHeaderLine getInfoLine(final String id) {
+        return (infoLines.containsKey(id) ? infoLines.get(id) : new VCFInfoHeaderLine(id, 1, VCFHeaderLineType.Integer, "(Key not found -- type may be incorrect)"));
```
This was the thing you helped me with a while ago with a totally perplexing HTSJDK header exception where the IntelliJ debugger doesn't output elements in a collection that are null, so the size and indexes look fine.  I wanted to put a placeholder because I never deal with the header minutiae until I'm ready to PR, but I can throw an exception.  I miss `--unsafe` mode.

ldgauthier commented:
@lbergelson I addressed all your comments.  I had some difficulty reconciling the list of Strings for annotation names to remove with the List<Annotation> for the getDefaultVariantAnnotations update, so there may be a few tests failing because the RAW_MQ header disappeared.  I'll deal with those when Travis is done, but could you take a look at the other changes?

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/ReblockGVCF.java:
```diff
@@ -0,0 +1,437 @@
+ * <pre>
+ * gatk ReblockGVCF \
+ *   -R reference.fasta \
+ *   --variant sample1.g.vcf \
```
I'd just make this -V to match the others.

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/ReblockGVCF.java:
```diff
@@ -0,0 +1,437 @@
+            doc="File to which variants should be written")
+    private File outputFile;
+
+    @Argument(fullName=GenotypeGVCFs.ALL_SITES_LONG_NAME, shortName=GenotypeGVCFs.ALL_SITES_SHORT_NAME, doc="Include loci found to be non-variant after genotyping")
```
It's a little weird to make a boolean argument default to true.  Should the name be inverted so that you specify it to disable it?

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/ReblockGVCF.java:
```diff
@@ -0,0 +1,437 @@
+    public GenotypeCalculationArgumentCollection genotypeArgs = new GenotypeCalculationArgumentCollection();
+
+    @Advanced
+    @Argument(fullName="GVCF-GQ-band", shortName="GQB", doc="Exclusive upper bounds for reference confidence GQ bands " +
```
this should use the new GQ_BAND_LONG_NAME  constants

lbergelson commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/ReblockGVCF.java:
```diff
@@ -0,0 +1,437 @@
+        try {
+            vcfWriter = new GVCFWriter(writer, GVCFGQBands, PLOIDY_TWO);
+        } catch ( IllegalArgumentException e ) {
+            throw new CommandLineException.BadArgumentValue("GQBands", "are malformed: " + e.getMessage());
```
Could you also propagate the underlying exception as the cause here?  It's always best to do.

lbergelson commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/variantutils/ReblockGVCFUnitTest.java:
```diff
@@ -0,0 +1,96 @@
+        Assert.assertTrue(!newG.hasAD());
+    }
+
+    //TODO: these are duplicated from PosteriorProbabilitiesUtilsUnitTest but PR #4947 modifies VariantContextTestUtils, so I'll do some refactoring before the second of the two is merged
```
We should remember to resolve this when #4947 is completed.

lbergelson commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/variantutils/ReblockGVCFUnitTest.java:
```diff
@@ -0,0 +1,96 @@
+    }
+
+    @Test
+    public void testLowQualVariantToGQ0HomRef() {
```
forgotten test?

lbergelson commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/variantutils/ReblockGVCFUnitTest.java:
```diff
@@ -0,0 +1,96 @@
+import java.util.Map;
+
+public class ReblockGVCFUnitTest {
+    private final ReblockGVCF reblocker = new ReblockGVCF();
```
I would just create this in each test that uses it, otherwise we potentially get weird stateful issues across multiple tests.

lbergelson commented on src/test/java/org/broadinstitute/hellbender/tools/walkers/variantutils/ReblockGVCFUnitTest.java:
```diff
@@ -0,0 +1,96 @@
+import java.util.List;
+import java.util.Map;
+
+public class ReblockGVCFUnitTest {
```
Thanks for adding these tests.  Complicated functions like these have a tendency to lose bits of logic later on if there aren't direct tests for them.

lbergelson commented on src/test/java/org/broadinstitute/hellbender/utils/variant/writers/GVCFWriterUnitTest.java:
```diff
@@ -629,4 +654,53 @@ public void testAgainstExampleGVCF() throws IOException {
+        final Allele alt1 = Allele.create("T", false);
+        final Allele ref2 = Allele.create("TACACACACACTACTA", true);
+        final Allele ref3 = Allele.create("C", true);
+        final VariantContext deletion1 = new VariantContextBuilder(null, "1", 10000, 10020, Arrays.asList(ref1, alt1,
```
setting these things up is always such a pain, thank you for writing it

ldgauthier commented:
@lbergelson this is good to go, unless my other PR gets merged first, in which case I'll need to update the expected test data.

lbergelson commented:
That person might have been me...

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/ReblockGVCF.java:
```diff
@@ -0,0 +1,430 @@
+
+    }
+
+    public void onTraversalStart() {
```
`@Override`

ldgauthier replied:
Done -- I don't know why github can't figure out that this has changed

droazen commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/ReblockGVCF.java:
```diff
@@ -0,0 +1,430 @@
+    }
+
+    // get VariantContexts from input gVCFs and regenotype
+    public void apply(VariantContext variant, ReadsContext reads, ReferenceContext ref, FeatureContext features) {
```
`@Override`

ldgauthier replied:
done

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/ReblockGVCF.java:
```diff
@@ -0,0 +1,430 @@
+ *
+ * <h3>Caveats</h3>
+ * <p>Only single-sample gVCF files produced by HaplotypeCaller can be used as input for this tool.</p>
+ * <p>By default this tool only passes through annotations used by VQSR.  A different set of annotations can be specified with the usual -A argument.  Annotation groups are not supported by this tool.</p>
```
I think annotation groups are automatically supported and propagated to the annotation engine via `makeVariantAnnotations`, unless the tool manually intervenes.

ldgauthier replied:
I agree -- that was probably just a leftover from the GATK3 version.

cmnbroad commented on src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/ReblockGVCF.java:
```diff
@@ -0,0 +1,430 @@
+
+        headerLines.addAll(getDefaultToolVCFHeaderLines());
+
+        genotypingEngine = createGenotypingEngine(new IndexedSampleList(inputHeader.getGenotypeSamples()));
```
What is the behavior if the input has multiple samples ? Should this verify that there is only a single sample in the input ?

lbergelson replied:
This seems like a good check to add.

lbergelson replied:
Oh, you did add it up above 👍

lbergelson commented:
@ldgauthier I think there's an newly introduce compile warning because of the method names that I deprecated in htsjdk.
